### PR TITLE
Node.js: Add example

### PR DESCRIPTION
**Pull request by davedoesdev**

### CODE REVIEW

**Comment by anttikantee** [comment]
Did you try a raw tap interface to see if the 5s delay is due to qemu's hostfwd or something else?

**Comment by davedoesdev** [comment]
Good idea

**Comment by davedoesdev** [comment]
Looks like it's hostfwd: bridge and tap networking is up straight away. Strange.

**Comment by davedoesdev** [comment]
If I do `-W nic,inet,static,10.0.2.15/24` (instead of `dhcp`), it's down to 2s, sometimes less.

**Comment by anttikantee** [comment]
Yea, dunno, I don't use hostfwd very much.  I'm 99% sure the delay is a qemu problem.

If you tcpdump the connect, does it take several seconds to connect, or several seconds for the response?  Or do you get ENOWHATEVER if you try to connect before the "timeout"?

**Comment by davedoesdev** [comment]
Actually, I can connect straight away after I get the "Express started on port 3000" message. The rough <2s was me connecting as it was booting.
Is there a dhcp client in guest?

**Comment by mato** [comment]
If you use `-W nic,inet,dhcp` when running the guest then you will get DHCP. Run `rumprun` with no arguments to get a list of supported options.

**Comment by davedoesdev** [comment]
So I see in `buildrump.sh/brlib/libnetconfig/dhcp_main.c` that it loops with sleep until it gets an address.
I guess QEMU's DHCP server could just be slow to give out an address.

**Comment by anttikantee** [comment]
Doubtful that the DHCP server is slow, at least for me the qemu internal one is instantaneous.  Anyway, you can easily observe it by eye by looking at the bootstrap output; the dhcp client prints the address it configures.

If you try to connect before the destination is up, I'd guess QEMU just drops the SYN from your host's stack, and you have to wait for your host to timeout and send another SYN before the connection can be established.  I don't think it's different from trying to connect to anything destination which is not up.

**Comment by anttikantee** [comment]
And just to clarify, by "up" I mean that the destination TCP/IP stack is not up, not just a port which is not listening.

**Comment by davedoesdev** [comment]
Would that be different for dhcp vs static though? When I configure a static address, I get a response straight away.

**Comment by anttikantee** [comment]
I don't know, you press "enter" at a slightly different time in the other terminal?  dhcp configuration will always take just a bit longer.

**Comment by davedoesdev** [comment]
Looks like it's duplicate address detection.
If I `return 0` from `if_do_dad` (`src-netbsd/sys/net/if.c`) then I get no delay for DHCP under KVM.

`-g '-net dump'` was my friend here.

**Comment by anttikantee** [comment]
Hmm.

Are you using the virtio NIC in all cases, or do you also use the emulated e1000 NIC?

**Comment by anttikantee** [comment]
Actually, can you summarize your different ways of running the guest (including full command line), and observed result?

**Comment by davedoesdev** [comment]
This won't reply to a HTTP request for 5 seconds. You can make a request but it doesn't get a reply until the VM has been up for 5 seconds.

``` shell
rumprun kvm -M 160 -g '-net dump' -I 'nic,vioif,-net user,hostfwd=tcp::3000-:3000' -W nic,inet,dhcp -i -b express-$(EXPRESS_VERSION).iso,/express ../build/out/Release/node.bin /express/examples/hello-world/index.js
```

The following replies straight away:

``` shell
rumprun kvm -M 160 -g '-net dump' -I 'nic,vioif,-net user,hostfwd=tcp::3000-:3000' -W nic,inet,static,10.0.2.15/24 -i -b express-$(EXPRESS_VERSION).iso,/express ../build/out/Release/node.bin /express/examples/hello-world/index.js
```

**Comment by anttikantee** [comment]
So the only difference is using dhcp or not.  Above you said that using a static address reduces the delay "down to 2s, sometimes less".  Was that statement incorrect?

**Comment by davedoesdev** [comment]
Yes: if I wait for Express to say it's listening on port 3000 then I always get an immediate response.
The 2s was me refreshing the browser too early.

**Comment by anttikantee** [comment]
Ok, thanks.  Can you open a new issue in rumprun for this problem?  I guess it's simply that doing dhcp config tickles DAD while static does not.  But I'll have a closer look some day.

**Comment by davedoesdev** [comment]
Sure, I'll open an issue tomorrow and attach pcaps if github will let me
